features/FEATURE_BLE/ble/gap/AdvertisingDataParser.h

```diff
@@ -0,0 +1,122 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BLE_GAP_ADVERTISINGDATAPARSER_H
+#define BLE_GAP_ADVERTISINGDATAPARSER_H
+
+#include <stdint.h>
+#include "ble/gap/AdvertisingDataTypes.h"
+#include "platform/Span.h"
+
+namespace ble {
+
+/**
+ * @addtogroup ble
+ * @{
+ * @addtogroup gap
+ * @{
+ */
+
+/**
+ * Parse and iterate over advertising data
+ */
+class AdvertisingDataParser {
+
+    enum {
+        DATA_SIZE_INDEX = 0,
+        TYPE_INDEX = 1,
+        VALUE_INDEX = 2,
+        TYPE_SIZE = 1,
+        DATA_SIZE_SIZE = 1
+    };
+
+public:
+    /**
+     * Representation of an Advertising Data element.
+     */
+    struct element_t {
+        adv_data_type_t type;
+        mbed::Span<const uint8_t> value;
+    };
+
+    /**
+     * Build a parser from an array of bytes.
+     * @param data The data to parse.
+     */
+    AdvertisingDataParser(mbed::Span<const uint8_t> data) :
+        data(data),
+        position(0)
+    {
+    }
+
+    /**
+     * Return if there is advertising data element left to parse.
+     */
+    bool hasNext() const
+    {
+        if (position >= data.size()) {
+            return false;
+        }
+
+        if (position + current_length() >= data.size()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Return the next advertising data element.
+     *
+     * @note Calling this function if there is no next element is undefined
```

Comment:
it needn't be, an empty element can be returned

Reply:
Yes it does, if you haven't called `hasNext` before `next` and there is no elements left it is like dereferencing an array without checking the bound.